pkg/service/one2onerestore/worker_validate.go

```diff
@@ -0,0 +1,213 @@
+	}
+
+	for _, target := range targetNodeInfo {
+		source, ok := mappedSourceNodeInfo[node{DC: target.DC, Rack: target.Rack, HostID: target.HostID}]
+		if !ok {
+			return errors.Errorf("target node has no match in source cluster:%s %s %s", target.DC, target.Rack, target.HostID)
+		}
+		if source.CPUCount != target.CPUCount {
+			return errors.Errorf("source CPUCount doesn't match target CPUCount")
+		}
+		if source.StorageSize > target.StorageSize {
+			return errors.Errorf("source StorageSize greater than target StorageSize")
+		}
+		if !slices.Equal(source.Tokens, target.Tokens) {
+			return errors.Errorf("source Tokens doesn't match target Tokens")
+		}
+	}
+	return nil
+}
+
+func mapTargetHostToSource(targetHosts []Host, nodeMappings []nodeMapping) (map[string]Host, error) {
+	sourceByTargetHostID := map[string]string{}
+	for _, mapping := range nodeMappings {
+		sourceByTargetHostID[mapping.Target.HostID] = mapping.Source.HostID
+	}
+
+	result := map[string]Host{}
+	for _, host := range targetHosts {
+		sourceHostID, ok := sourceByTargetHostID[host.ID]
+		if !ok {
+			return nil, errors.Errorf("mapping for target node (%s) is not found", host.ID)
+		}
+		result[sourceHostID] = host
+	}
+	return result, nil
+}
+
+func mapSourceNodesToTarget(sourceNodeInfo []nodeValidationInfo, nodeMappings []nodeMapping) (map[node]nodeValidationInfo, error) {
+	sourceMappings := map[node]node{}
+	for _, m := range nodeMappings {
+		sourceMappings[m.Source] = m.Target
+	}
+
+	sourceByTarget := map[node]nodeValidationInfo{}
+	for _, source := range sourceNodeInfo {
+		target, ok := sourceMappings[node{DC: source.DC, Rack: source.Rack, HostID: source.HostID}]
+		if !ok {
+			return nil, errors.Errorf("mapping for source node (%v) is not found", source)
+		}
+		sourceByTarget[target] = source
+	}
+	return sourceByTarget, nil
+}
+
+func (w *worker) collectNodeValidationInfo(
+	ctx context.Context,
+	manifests []*backupspec.ManifestInfo,
+	hosts []Host,
+	nodeMappings []nodeMapping,
+) (sourceCluster, targetCluster []nodeValidationInfo, err error) {
+	targetHostBySourceID, err := mapTargetHostToSource(hosts, nodeMappings)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "invalid node mappings")
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		sourceInfo, err := w.getSourceClusterNodeInfo(ctx, manifests, targetHostBySourceID)
+		if err != nil {
+			return errors.Wrap(err, "source cluster node info")
+		}
+
+		sourceCluster = sourceInfo
+		return nil
+	})
+	g.Go(func() error {
+		targetInfo, err := w.getTargetClusterNodeInfo(ctx, hosts)
+		if err != nil {
+			return errors.Wrap(err, "target cluster node info")
+		}
+
+		targetCluster = targetInfo
+		return nil
+	})
+
+	if err := g.Wait(); err != nil {
+		return nil, nil, err
+	}
+
+	return sourceCluster, targetCluster, nil
+}
+
+type nodeValidationInfo struct {
+	DC          string
+	Rack        string
+	HostID      string
+	CPUCount    int64
```

Comment:
Shouldn't it be shard count instead of cpu count?

Reply:
Agree, it should be shards.
CPU defines amount of available CPUs on the machine.
Shards are defining how many CPUs are used by Scylla.